# **Aprendizagem Supervisionada: Classificação**

Desenvolvimento de um algoritmo de Machine Learning para prever a tendência de uma pessoa desenvolver algum tipo de doença cardíaca com base em alguns fatores clínicos e laboratoriais.

Os dados foram extraídos do Kaggle:

https://www.kaggle.com/fedesoriano/heart-failure-prediction/version/1

# **Pré-Processamento**

In [123]:
import numpy as np
import pandas as pd

In [124]:
df = pd.read_csv('/content/drive/MyDrive/curso/heart_tratado.csv', sep=';', encoding='iso-8859-1')   # ou utf-8

In [125]:
df.head()       # cabeçalho da tabela de dados

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140        289.0          0     Normal    172   
1   49   F           NAP        160        180.0          0     Normal    156   
2   37   M           ATA        130        283.0          0         ST     98   
3   48   F           ASY        138        214.0          0     Normal    108   
4   54   M           NAP        150        195.0          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [126]:
df.shape    # indica a quantidade de linhas e colunas

(917, 12)

# **Transformando as variáveis categóricas nominais em variáveis categóricas ordinais**

In [127]:
df2 = pd.DataFrame.copy(df)   # utilizando uma cópia do DataFrame anterior

In [128]:
df2['Sex'].replace({'M':0, 'F':1}, inplace=True)
df2['ChestPainType'].replace({'TA':0, 'ATA':1,  'NAP':2, 'ASY':3}, inplace=True)
df2['RestingECG'].replace({'Normal':0, 'ST':1, 'LVH':2}, inplace=True)
df2['ExerciseAngina'].replace({'N':0, 'Y':1}, inplace=True)
df2['ST_Slope'].replace({'Up':0, 'Flat':1, 'Down':2}, inplace=True)

In [129]:
df2.head()               # cabeçalho da tabela de dados

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40    0              1        140        289.0          0           0   
1   49    1              2        160        180.0          0           0   
2   37    0              1        130        283.0          0           1   
3   48    1              3        138        214.0          0           0   
4   54    0              2        150        195.0          0           0   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         0             0  
1    156               0      1.0         1             1  
2     98               0      0.0         0             0  
3    108               1      1.5         1             1  
4    122               0      0.0         0             0

In [130]:
df2.dtypes                       # informa o tipo de todos os atributos

Age                 int64
Sex                 int64
ChestPainType       int64
RestingBP           int64
Cholesterol       float64
FastingBS           int64
RestingECG          int64
MaxHR               int64
ExerciseAngina      int64
Oldpeak           float64
ST_Slope            int64
HeartDisease        int64
dtype: object

In [131]:
df2.shape    # indica a quantidade de linhas e colunas

(917, 12)

# **Legenda**

Age = idade (anos)

Sex = sexo (0=M; 1=F)

Chest Pain Type = tipo de dor no peito (0=TA:angina típica; 1=ATA:angina atípica; 2=NAP:dor não anginosa; 3=ASY:assintomático)

Resting BP = pressão sanguínea em repouso (mmHg)

Cholesterol = colesterol sérico (mg/dl)

Fasting BS = açúcar no sangue em jejum (mg/dl) 0: Fasting BS < 120 mg/dl (não diabético) 1: Fasting BS >= 120 mg/dl, (diabético)

Resting ECG = eletrocardiograma em repouso (0=Normal; 1=ST:Anormalidade da onda ST-T; 2=LVH: Hipertrofia ventricular esquerda)

Max HR = frequência cardíaca máxima

Exercise Angina = Angina induzida por exercício (0=Não; 1=Sim)

Old Peak = Depressão de ST induzida por exercício em relação ao repouso

ST_Slope = Inclinação do segmento ST (0=UP; 1=Flat; 2=Down)

Heart Disease = Doença cardíaca (0=Não possui doença cardíaca; 1=Possui doença cardíaca)

# **Atributos Previsores e Alvo**

In [132]:
df2.head()           # cabeçalho da tabela de dados

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40    0              1        140        289.0          0           0   
1   49    1              2        160        180.0          0           0   
2   37    0              1        130        283.0          0           1   
3   48    1              3        138        214.0          0           0   
4   54    0              2        150        195.0          0           0   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         0             0  
1    156               0      1.0         1             1  
2     98               0      0.0         0             0  
3    108               1      1.5         1             1  
4    122               0      0.0         0             0

In [133]:
previsores = df2.iloc[:, 0:11].values   # localizando os valores linhas e colunas de tendência pelo índice em formato de matriz

In [134]:
previsores    # verificando a matriz

array([[40. ,  0. ,  1. , ...,  0. ,  0. ,  0. ],
       [49. ,  1. ,  2. , ...,  0. ,  1. ,  1. ],
       [37. ,  0. ,  1. , ...,  0. ,  0. ,  0. ],
       ...,
       [57. ,  0. ,  3. , ...,  1. ,  1.2,  1. ],
       [57. ,  1. ,  1. , ...,  0. ,  0. ,  1. ],
       [38. ,  0. ,  2. , ...,  0. ,  0. ,  0. ]])

In [135]:
previsores.shape      # indica a quantidade de linhas e colunas

(917, 11)

In [136]:
alvo = df.iloc[:, 11].values      # selecionando todas as linhas da coluna 11 referente ao alvo

In [137]:
alvo          # verificando a matriz alvo separado dos previsores

array([0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,

In [138]:
alvo.shape            # indica a quantidade de linhas e colunas

(917,)

# **Análise das Escalas dos Atributos (Escalonamento)**

In [139]:
df2.describe()  # apresentando as estatísticas

Age         Sex  ChestPainType   RestingBP  Cholesterol  \
count  917.000000  917.000000     917.000000  917.000000   917.000000   
mean    53.509269    0.210469       2.251908  132.540894   244.635389   
std      9.437636    0.407864       0.931502   17.999749    53.347125   
min     28.000000    0.000000       0.000000   80.000000    85.000000   
25%     47.000000    0.000000       2.000000  120.000000   214.000000   
50%     54.000000    0.000000       3.000000  130.000000   244.635389   
75%     60.000000    0.000000       3.000000  140.000000   267.000000   
max     77.000000    1.000000       3.000000  200.000000   603.000000   

        FastingBS  RestingECG       MaxHR  ExerciseAngina     Oldpeak  \
count  917.000000  917.000000  917.000000      917.000000  917.000000   
mean     0.233370    0.604144  136.789531        0.404580    0.886696   
std      0.423206    0.806161   25.467129        0.491078    1.066960   
min      0.000000    0.000000   60.000000        0.000000   -2.600000   
25%      0.000000    0.000000  120.000000        0.000000    0.000000   
50%      0.000000    0.000000  138.000000        0.000000    0.600000   
75%      0.000000    1.000000  156.000000        1.000000    1.500000   
max      1.000000    2.000000  202.000000        1.000000    6.200000   

        ST_Slope  HeartDisease  
count  917.00000    917.000000  
mean     0.63795      0.552890  
std      0.60727      0.497466  
min      0.00000      0.000000  
25%      0.00000      0.000000  
50%      1.00000      1.000000  
75%      1.00000      1.000000  
max      2.00000      1.000000

Padronização (utiliza a média e o desvio padrão como referência).

Normalização (utiliza os valores máximo e mínimo como referência).

In [140]:
from sklearn.preprocessing import StandardScaler    # escala de padronização

In [141]:
previsores_esc = StandardScaler().fit_transform(previsores)   # todos os valores do selecionados para 'previsores' exceto a última coluna  'alvo'
# realizando uma transformação em previsores em relação à escala

In [142]:
previsores_esc     # transformando os valores com uma média próxima de 0 com desvio padrão próximo de 1

array([[-1.43220634, -0.51630861, -1.34470119, ..., -0.82431012,
        -0.83150225, -1.05109458],
       [-0.47805725,  1.9368261 , -0.27058012, ..., -0.82431012,
         0.10625149,  0.59651863],
       [-1.75025603, -0.51630861, -1.34470119, ..., -0.82431012,
        -0.83150225, -1.05109458],
       ...,
       [ 0.37007527, -0.51630861,  0.80354095, ...,  1.21313565,
         0.29380223,  0.59651863],
       [ 0.37007527,  1.9368261 , -1.34470119, ..., -0.82431012,
        -0.83150225,  0.59651863],
       [-1.64423947, -0.51630861, -0.27058012, ..., -0.82431012,
        -0.83150225, -1.05109458]])

In [143]:
previsoresdf = pd.DataFrame(previsores_esc)     # transformando os valores númericos 'padronizados' em DataFrame
previsoresdf

0         1         2         3         4         5         6   \
0   -1.432206 -0.516309 -1.344701  0.414627  0.832075 -0.551733 -0.749818   
1   -0.478057  1.936826 -0.270580  1.526360 -1.212261 -0.551733 -0.749818   
2   -1.750256 -0.516309 -1.344701 -0.141240  0.719543 -0.551733  0.491306   
3   -0.584074  1.936826  0.803541  0.303453 -0.574578 -0.551733 -0.749818   
4    0.052026 -0.516309 -0.270580  0.970493 -0.930931 -0.551733 -0.749818   
..        ...       ...       ...       ...       ...       ...       ...   
912 -0.902124 -0.516309 -2.418822 -1.252973  0.363191 -0.551733 -0.749818   
913  1.536257 -0.516309  0.803541  0.636973 -0.968441  1.812470 -0.749818   
914  0.370075 -0.516309  0.803541 -0.141240 -2.131275 -0.551733 -0.749818   
915  0.370075  1.936826 -1.344701 -0.141240 -0.161960 -0.551733  1.732430   
916 -1.644239 -0.516309 -0.270580  0.303453 -1.306038 -0.551733 -0.749818   

           7         8         9         10  
0    1.383339 -0.824310 -0.831502 -1.051095  
1    0.754736 -0.824310  0.106251  0.596519  
2   -1.523953 -0.824310 -0.831502 -1.051095  
3   -1.131075  1.213136  0.575128  0.596519  
4   -0.581047 -0.824310 -0.831502 -1.051095  
..        ...       ...       ...       ...  
912 -0.188170 -0.824310  0.293802  0.596519  
913  0.165420 -0.824310  2.356860  0.596519  
914 -0.856061  1.213136  0.293802  0.596519  
915  1.461915 -0.824310 -0.831502  0.596519  
916  1.422627 -0.824310 -0.831502 -1.051095  

[917 rows x 11 columns]

In [144]:
previsoresdf.describe()   # apresentando as estatísticas

0             1             2             3             4   \
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02   
mean   1.859654e-16  7.748558e-18  1.046055e-16  7.767929e-16 -1.869340e-16   
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00   
min   -2.704405e+00 -5.163086e-01 -2.418822e+00 -2.920572e+00 -2.994023e+00   
25%   -6.900904e-01 -5.163086e-01 -2.705801e-01 -6.971063e-01 -5.745784e-01   
50%    5.202558e-02 -5.163086e-01  8.035410e-01 -1.412398e-01  0.000000e+00   
75%    6.881250e-01 -5.163086e-01  8.035410e-01  4.146267e-01  4.194568e-01   
max    2.490407e+00  1.936826e+00  8.035410e-01  3.749826e+00  6.721265e+00   

                 5           6             7             8             9   \
count  9.170000e+02  917.000000  9.170000e+02  9.170000e+02  9.170000e+02   
mean   4.649135e-17    0.000000 -5.114048e-16 -1.046055e-16  7.748558e-17   
std    1.000546e+00    1.000546  1.000546e+00  1.000546e+00  1.000546e+00   
min   -5.517333e-01   -0.749818 -3.016886e+00 -8.243101e-01 -3.269662e+00   
25%   -5.517333e-01   -0.749818 -6.596226e-01 -8.243101e-01 -8.315022e-01   
50%   -5.517333e-01   -0.749818  4.755658e-02 -8.243101e-01 -2.688500e-01   
75%   -5.517333e-01    0.491306  7.547357e-01  1.213136e+00  5.751284e-01   
max    1.812470e+00    1.732430  2.561971e+00  1.213136e+00  4.982571e+00   

                 10  
count  9.170000e+02  
mean  -3.874279e-17  
std    1.000546e+00  
min   -1.051095e+00  
25%   -1.051095e+00  
50%    5.965186e-01  
75%    5.965186e-01  
max    2.244132e+00

# **Codificação de Variáveis Categóricas Automaticamente**

## **LabelEncoder: Transformação de variáveis categóricas em númericas**

In [145]:
from sklearn.preprocessing import LabelEncoder   # transformar variáveis categóricas em númericas ou ordinais

In [146]:
df.head()           # cabeçalho da tabela de dados

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140        289.0          0     Normal    172   
1   49   F           NAP        160        180.0          0     Normal    156   
2   37   M           ATA        130        283.0          0         ST     98   
3   48   F           ASY        138        214.0          0     Normal    108   
4   54   M           NAP        150        195.0          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [147]:
previsores2 = df.iloc[:, 0:11].values           # separando de 0 a 11 os valores 'previsores não adaptados' do 'alvo não adaptado'
previsores2    # apresentando a configuração no formato de matriz

array([[40, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       [49, 'F', 'NAP', ..., 'N', 1.0, 'Flat'],
       [37, 'M', 'ATA', ..., 'N', 0.0, 'Up'],
       ...,
       [57, 'M', 'ASY', ..., 'Y', 1.2, 'Flat'],
       [57, 'F', 'ATA', ..., 'N', 0.0, 'Flat'],
       [38, 'M', 'NAP', ..., 'N', 0.0, 'Up']], dtype=object)

In [148]:
previsores2[:,1] = LabelEncoder().fit_transform(previsores2[:,1])   # transformando 'padronizando' todas as linhas da coluna 1

In [149]:
previsores2  # apresentando a padronização de todas as linhas da coluna 1

array([[40, 1, 'ATA', ..., 'N', 0.0, 'Up'],
       [49, 0, 'NAP', ..., 'N', 1.0, 'Flat'],
       [37, 1, 'ATA', ..., 'N', 0.0, 'Up'],
       ...,
       [57, 1, 'ASY', ..., 'Y', 1.2, 'Flat'],
       [57, 0, 'ATA', ..., 'N', 0.0, 'Flat'],
       [38, 1, 'NAP', ..., 'N', 0.0, 'Up']], dtype=object)

In [150]:
previsores2[:,2] = LabelEncoder().fit_transform(previsores2[:,2])   # transformando 'padronizando' todas as linhas da coluna 2
previsores2[:,6] = LabelEncoder().fit_transform(previsores2[:,6])   # transformando 'padronizando' todas as linhas da coluna 6
previsores2[:,8] = LabelEncoder().fit_transform(previsores2[:,8])   # transformando 'padronizando' todas as linhas da coluna 8
previsores2[:,10] = LabelEncoder().fit_transform(previsores2[:,10])   # transformando 'padronizando' todas as linhas da coluna 10

In [151]:
previsores2   # apresentando a padronização de todas as linhas e colunas que precisavam de transformação númerica

array([[40, 1, 1, ..., 0, 0.0, 2],
       [49, 0, 2, ..., 0, 1.0, 1],
       [37, 1, 1, ..., 0, 0.0, 2],
       ...,
       [57, 1, 0, ..., 1, 1.2, 1],
       [57, 0, 1, ..., 0, 0.0, 1],
       [38, 1, 2, ..., 0, 0.0, 2]], dtype=object)

In [152]:
previsores2.shape            # indica a quantidade de linhas e colunas

(917, 11)

# **OneHotEncoder: Criação de variáveis Dummy (fictícia)**

Variáveis criadas para evitar a existência de outras variáveis altamente correlacionadas entre si.

A = 0, B = 1, C = 2, D = 3


A   B   C   D   
1   0   0   0   
0   1   0   0   
0   0   1   0   
0   0   0   1


In [153]:
from sklearn.preprocessing import OneHotEncoder
from sklearn. compose import ColumnTransformer

In [154]:
ColumnTransformer?

Parâmetros ColumnTransformer
- name: nome dado a transformação.
- transformer: tipo de estimador (OneHotEncoder).
- columns: colunas que serão transformadas.
- remainder: o que acontecerá com o restante das colunas não relacionadas:
1) drop = exclui as outras colunas.
2) passthrough = mantém as outras colunas. drop é default.  
- sparse_threshold: parâmetro de classificação de matrizes esparsas. default é 0.3
- n_jobs: número de trabalhos a serem executados em paralelo. default é nenhum
- transformer_weights: definição de pesos aos transformadores.
-verbose: default é False. se for True a execução é apresentada na tela.

In [155]:
previsores3 = ColumnTransformer(transformers=[('OneHot', OneHotEncoder(), [1,2,6,8,10])],
                                remainder='passthrough').fit_transform(previsores2)
# estabelecendo o nome da transformação, tipo de transformação, colunas que serão transformadas, colunas que serão mantidas ou excluídas.

In [156]:
previsores3               # apresentando a padronização de todas as linhas e colunas que precisavam de transformação númerica

array([[0.0, 1.0, 0.0, ..., 0, 172, 0.0],
       [1.0, 0.0, 0.0, ..., 0, 156, 1.0],
       [0.0, 1.0, 0.0, ..., 0, 98, 0.0],
       ...,
       [0.0, 1.0, 1.0, ..., 0, 115, 1.2],
       [1.0, 0.0, 0.0, ..., 0, 174, 0.0],
       [0.0, 1.0, 0.0, ..., 0, 173, 0.0]], dtype=object)

In [157]:
previsores3.shape               # indica a quantidade de linhas e colunas
# foram adicionadas novos atributos ou variáveis em colunas através da transformação 11 para 20 colunas

(917, 20)

In [158]:
previsores3df = pd.DataFrame(previsores3)             # transformando os valores numéricos 'padronizados' em DataFrame
previsores3df.head()

# LabelEncoder transformando as variáveis nominais em númericas
# OneHotEncoder criando as variáveis dummy

0    1    2    3    4    5    6    7    8    9    10   11   12   13  14  \
0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  40   
1  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0  49   
2  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  1.0  37   
3  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  48   
4  0.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  54   

    15     16 17   18   19  
0  140  289.0  0  172  0.0  
1  160  180.0  0  156  1.0  
2  130  283.0  0   98  0.0  
3  138  214.0  0  108  1.5  
4  150  195.0  0  122  0.0

In [159]:
df.head()             # cabeçalho da tabela de dados

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140        289.0          0     Normal    172   
1   49   F           NAP        160        180.0          0     Normal    156   
2   37   M           ATA        130        283.0          0         ST     98   
3   48   F           ASY        138        214.0          0     Normal    108   
4   54   M           NAP        150        195.0          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

# **Escalonamento**

In [160]:
from sklearn.preprocessing import StandardScaler    # escala de padronização

In [161]:
previsores3_esc = StandardScaler().fit_transform(previsores3)    # todos os valores do selecionados para escalonamento de 'previsores3'
# realizando uma transformação em previsores em relação à escala

In [162]:
previsores3_esc     # transformando os valores com uma média próxima de 0 com desvio padrão próximo de 1

array([[-0.51630861,  0.51630861, -1.08542493, ..., -0.55173333,
         1.38333943, -0.83150225],
       [ 1.9368261 , -1.9368261 , -1.08542493, ..., -0.55173333,
         0.75473573,  0.10625149],
       [-0.51630861,  0.51630861, -1.08542493, ..., -0.55173333,
        -1.52395266, -0.83150225],
       ...,
       [-0.51630861,  0.51630861,  0.92129817, ..., -0.55173333,
        -0.85606123,  0.29380223],
       [ 1.9368261 , -1.9368261 , -1.08542493, ..., -0.55173333,
         1.46191489, -0.83150225],
       [-0.51630861,  0.51630861, -1.08542493, ..., -0.55173333,
         1.42262716, -0.83150225]])

In [163]:
previsores3df = pd.DataFrame(previsores3_esc)   # transformando os valores com uma média próxima de 0 com desvio padrão próximo de 1
previsores3df

0         1         2         3         4         5         6   \
0   -0.516309  0.516309 -1.085425  2.073784 -0.531524 -0.229810 -0.507826   
1    1.936826 -1.936826 -1.085425 -0.482210  1.881384 -0.229810 -0.507826   
2   -0.516309  0.516309 -1.085425  2.073784 -0.531524 -0.229810 -0.507826   
3    1.936826 -1.936826  0.921298 -0.482210 -0.531524 -0.229810 -0.507826   
4   -0.516309  0.516309 -1.085425 -0.482210  1.881384 -0.229810 -0.507826   
..        ...       ...       ...       ...       ...       ...       ...   
912 -0.516309  0.516309 -1.085425 -0.482210 -0.531524  4.351412 -0.507826   
913 -0.516309  0.516309  0.921298 -0.482210 -0.531524 -0.229810 -0.507826   
914 -0.516309  0.516309  0.921298 -0.482210 -0.531524 -0.229810 -0.507826   
915  1.936826 -1.936826 -1.085425  2.073784 -0.531524 -0.229810  1.969177   
916 -0.516309  0.516309 -1.085425 -0.482210  1.881384 -0.229810 -0.507826   

           7         8         9         10        11        12        13  \
0    0.815013 -0.490781  0.824310 -0.824310 -0.271607 -1.001091  1.149573   
1    0.815013 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
2   -1.226974  2.037569  0.824310 -0.824310 -0.271607 -1.001091  1.149573   
3    0.815013 -0.490781 -1.213136  1.213136 -0.271607  0.998910 -0.869888   
4    0.815013 -0.490781  0.824310 -0.824310 -0.271607 -1.001091  1.149573   
..        ...       ...       ...       ...       ...       ...       ...   
912  0.815013 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
913  0.815013 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
914  0.815013 -0.490781 -1.213136  1.213136 -0.271607  0.998910 -0.869888   
915 -1.226974 -0.490781  0.824310 -0.824310 -0.271607  0.998910 -0.869888   
916  0.815013 -0.490781  0.824310 -0.824310 -0.271607 -1.001091  1.149573   

           14        15        16        17        18        19  
0   -1.432206  0.414627  0.832075 -0.551733  1.383339 -0.831502  
1   -0.478057  1.526360 -1.212261 -0.551733  0.754736  0.106251  
2   -1.750256 -0.141240  0.719543 -0.551733 -1.523953 -0.831502  
3   -0.584074  0.303453 -0.574578 -0.551733 -1.131075  0.575128  
4    0.052026  0.970493 -0.930931 -0.551733 -0.581047 -0.831502  
..        ...       ...       ...       ...       ...       ...  
912 -0.902124 -1.252973  0.363191 -0.551733 -0.188170  0.293802  
913  1.536257  0.636973 -0.968441  1.812470  0.165420  2.356860  
914  0.370075 -0.141240 -2.131275 -0.551733 -0.856061  0.293802  
915  0.370075 -0.141240 -0.161960 -0.551733  1.461915 -0.831502  
916 -1.644239  0.303453 -1.306038 -0.551733  1.422627 -0.831502  

[917 rows x 20 columns]

In [164]:
previsores3df.describe()           # apresentando as estatísticas

0             1             2             3             4   \
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02   
mean   1.084798e-16 -1.472226e-16  1.937139e-17 -3.874279e-17  3.874279e-17   
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00   
min   -5.163086e-01 -1.936826e+00 -1.085425e+00 -4.822104e-01 -5.315237e-01   
25%   -5.163086e-01  5.163086e-01 -1.085425e+00 -4.822104e-01 -5.315237e-01   
50%   -5.163086e-01  5.163086e-01  9.212982e-01 -4.822104e-01 -5.315237e-01   
75%   -5.163086e-01  5.163086e-01  9.212982e-01 -4.822104e-01 -5.315237e-01   
max    1.936826e+00  5.163086e-01  9.212982e-01  2.073784e+00  1.881384e+00   

                 5           6             7             8             9   \
count  9.170000e+02  917.000000  9.170000e+02  9.170000e+02  9.170000e+02   
mean   6.973702e-17    0.000000 -9.298269e-17  1.549712e-17 -4.261707e-17   
std    1.000546e+00    1.000546  1.000546e+00  1.000546e+00  1.000546e+00   
min   -2.298105e-01   -0.507826 -1.226974e+00 -4.907810e-01 -1.213136e+00   
25%   -2.298105e-01   -0.507826 -1.226974e+00 -4.907810e-01 -1.213136e+00   
50%   -2.298105e-01   -0.507826  8.150134e-01 -4.907810e-01  8.243101e-01   
75%   -2.298105e-01   -0.507826  8.150134e-01 -4.907810e-01  8.243101e-01   
max    4.351412e+00    1.969177  8.150134e-01  2.037569e+00  8.243101e-01   

                 10            11          12            13            14  \
count  9.170000e+02  9.170000e+02  917.000000  9.170000e+02  9.170000e+02   
mean   4.261707e-17  8.523413e-17    0.000000 -3.874279e-17  1.859654e-16   
std    1.000546e+00  1.000546e+00    1.000546  1.000546e+00  1.000546e+00   
min   -8.243101e-01 -2.716072e-01   -1.001091 -8.698879e-01 -2.704405e+00   
25%   -8.243101e-01 -2.716072e-01   -1.001091 -8.698879e-01 -6.900904e-01   
50%   -8.243101e-01 -2.716072e-01    0.998910 -8.698879e-01  5.202558e-02   
75%    1.213136e+00 -2.716072e-01    0.998910  1.149573e+00  6.881250e-01   
max    1.213136e+00  3.681787e+00    0.998910  1.149573e+00  2.490407e+00   

                 15            16            17            18            19  
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  
mean   7.884157e-16  3.014189e-15 -1.549712e-17 -5.114048e-16 -1.859654e-16  
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  
min   -2.920572e+00 -2.994023e+00 -5.517333e-01 -3.016886e+00 -3.269662e+00  
25%   -6.971063e-01 -5.745784e-01 -5.517333e-01 -6.596226e-01 -8.315022e-01  
50%   -1.412398e-01  3.198360e-15 -5.517333e-01  4.755658e-02 -2.688500e-01  
75%    4.146267e-01  4.194568e-01 -5.517333e-01  7.547357e-01  5.751284e-01  
max    3.749826e+00  6.721265e+00  1.812470e+00  2.561971e+00  4.982571e+00

## **RESUMO PRÉ-PROCESSAMENTO**

alvo = variável que se pretende atingir (tem ou não doença cardíaca).

previsores = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas manualmente, sem escalonar.

previsores_esc = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas, escalonada.

previsores2 = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas pelo labelencoder.

previsores3 = conjunto de variáveis previsoras transformadas pelo labelencoder e onehotencoder, sem escalonar.

previsores3_esc = conjunto de variáveis previsoras transformadas pelo labelencoder e onehotencoder escalonada.

# **Base de Treino e Teste**

In [165]:
from sklearn.model_selection import train_test_split

Parâmetros train_test_split:   
- arrays: nomes dos atributos previsores e alvo.   
- test_size: tamanho em porcentagem dos dados de teste. default é none.   
- train_size: tamanho em porcentagem dos dados de treinamento.default é none.  
- random_state: nomeação de um estado aleatório.   
- shuffle: embaralhamento dos dados aleatórios. Associado com o random_state ocorre o mesmo embaralhamento sempre. Default é True.  
- stratify: Possibilidade de dividir os dados de forma estratificada. Default é None (nesse caso é mantido a proporção, isto é, se tem 30% de zeros e 70% de 1 no dataframe, na separação em treinamento e teste se manterá essa proporção).

In [166]:
x_treino, x_teste, y_treino, y_teste = train_test_split(previsores, alvo, test_size = 0.3, random_state = 0)
# x_treino e x_teste refere-se aos previsores, y_treino e y_teste refere-se ao alvo

In [167]:
x_treino.shape

(641, 11)

In [168]:
x_teste.shape

(276, 11)

In [169]:
y_treino.shape

(641,)

In [170]:
y_teste.shape

(276,)

# **XGBOOST**

https://xgboost.readthedocs.io/en/stable/

In [171]:
from xgboost import XGBClassifier

In [172]:
xg = XGBClassifier(max_depth=2, learning_rate=0.05, n_estimators=250, objective='binary:logistic', random_state=3)
xg.fit(x_treino,y_treino)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.05, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=2, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=250, n_jobs=None,
              num_parallel_tree=None, random_state=3, ...)

In [173]:
previsoes_xg = xg.predict(x_teste)
previsoes_xg

array([1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1])

In [174]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [175]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_xg) * 100.0))

Acurácia: 87.32%


In [176]:
confusion_matrix(y_teste, previsoes_xg)

array([[104,  17],
       [ 18, 137]])

In [177]:
print(classification_report(y_teste, previsoes_xg))

              precision    recall  f1-score   support

           0       0.85      0.86      0.86       121
           1       0.89      0.88      0.89       155

    accuracy                           0.87       276
   macro avg       0.87      0.87      0.87       276
weighted avg       0.87      0.87      0.87       276



**Análise dados de treino**

In [178]:
previsoes_treino = xg.predict(x_treino)
previsoes_treino

array([1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0,
       1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1,
       0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1,

In [179]:
accuracy_score(y_treino, previsoes_treino)

0.9173166926677067

In [180]:
confusion_matrix(y_treino, previsoes_treino)

array([[255,  34],
       [ 19, 333]])

### **Validação Cruzada**

In [181]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [182]:
# Separando os dados em folds
kfold = KFold(n_splits = 30, shuffle=True, random_state = 5)

In [186]:
# Criando o modelo
modelo = XGBClassifier(max_depth=2, learning_rate=0.05, n_estimators=250, objective='binary:logistic', random_state=3)
resultado = cross_val_score(modelo, previsores, alvo, cv = kfold)

# Usamos a média e o desvio padrão
print("Acurácia Média: %.2f%%" % (resultado.mean() * 100.0))

Acurácia Média: 87.02%


Naive Bayes = 84,78% (treino e teste) - 234 acertos e 85,17% (validação cruzada) : previsores3_esc

SVM = 86,23% - 238 acertos e 85,72% (validação cruzada) : previsores3-esc - SVC(kernel='rbf', random_state=1, C = 2)

Regressão logística = 86,23% - 238 acertos e 85,83% (validação cruzada) : previsores3_esc - LogisticRegression(random_state=1, max_iter=600, penalty="l2", tol=0.0001, C=1,solver="lbfgs")

KNN = 84,06% - 232 acertos e 85,84% (validação cruzada) : previsores3_esc ou previsores_esc - KNeighborsClassifier(n_neighbors=7, metric='minkowski', p = 1)

Árvore de decisão = 83,7% (treino e teste) - 231 acertos e 83,32% (validação cruzada): previsores3_esc - DecisionTreeClassifier(criterion='entropy', random_state = 0, max_depth=3)

Random Forest = 86,59% - 239 acertos e 85,92% (validação cruzada) : previsores - RandomForestClassifier(n_estimators=150, criterion='entropy', random_state = 0, max_depth=4)

XGboost = 87,68% - 242 acertos e 87,13% (validação cruzada) : previsores - XGBClassifier(max_depth=2, learning_rate=0.05, n_estimators=250, objective='binary:logistic', random_state=3)

# **LIGHTGBM**